plugins/catalog-backend/src/next/CatalogProcessingOrchestratorImpl.ts

```diff
@@ -0,0 +1,318 @@
+    if (!locationRef) {
+      throw new InputError(`Entity '${entityRef}' does not have a location`);
+    }
+    const location = parseLocationReference(locationRef);
+    const originLocation = parseLocationReference(
+      getEntityOriginLocationRef(unprocessedEntity),
+    );
+
+    const emitter = createEmitter(this.options.logger, unprocessedEntity);
+    try {
+      // Pre-process phase, used to populate entities with data that is required during main processing step
+      let entity = unprocessedEntity;
+      for (const processor of this.options.processors) {
+        if (processor.preProcessEntity) {
+          try {
+            entity = await processor.preProcessEntity(
+              entity,
+              location,
+              emitter.emit,
+              originLocation,
+            );
+          } catch (e) {
+            throw new Error(
+              `Processor ${processor.constructor.name} threw an error while preprocessing entity ${entityRef} at ${locationRef}, ${e}`,
+            );
+          }
+        }
+      }
+
+      // Enforce entity policies making sure that entities conform to a general schema
+      let policyEnforcedEntity;
+      try {
+        policyEnforcedEntity = await this.options.policy.enforce(entity);
+      } catch (e) {
+        throw new InputError(
+          `Policy check failed while analyzing entity ${entityRef} at ${locationRef}, ${e}`,
+        );
+      }
+      if (!policyEnforcedEntity) {
+        throw new Error(
+          `Policy unexpectedly returned no data while analyzing entity ${entityRef} at ${locationRef}`,
+        );
+      }
+      entity = policyEnforcedEntity;
+
+      // Validate the given entity kind against its schema
+      let handled = false;
+      for (const processor of this.options.processors) {
+        if (processor.validateEntityKind) {
+          try {
+            handled = await processor.validateEntityKind(entity);
+            if (handled) {
+              break;
+            }
+          } catch (e) {
+            throw new InputError(
+              `Processor ${processor.constructor.name} threw an error while validating the entity ${entityRef} at ${locationRef}, ${e}`,
+            );
+          }
+        }
+      }
+      if (!handled) {
+        throw new InputError(
+          `No processor recognized the entity ${entityRef} at ${locationRef}`,
+        );
+      }
+
+      // Backwards compatible processing of location entites
+      if (isLocationEntity(entity)) {
+        const { type = location.type } = entity.spec;
+        const targets = new Array<string>();
+        if (entity.spec.target) {
+          targets.push(entity.spec.target);
+        }
+        if (entity.spec.targets) {
+          targets.push(...entity.spec.targets);
+        }
+
+        for (const maybeRelativeTarget of targets) {
+          if (type === 'file' && maybeRelativeTarget.endsWith(path.sep)) {
```

Comment:
Good question, I would imagine entity policy can only apply to the data that is know by the entity schema? The fact that `file` type locations exist is not part of the schema, as it just defined `type:target`. With that you ofc couldn't enforce any policy on the `target` format either. Technically we could add the `file` type to be part of the schema, but we can't add all types since you can define custom ones, and we'd then end up with validation code in multiple different layers instead.

This particular validation here is just because we have to resolve relative paths at this point, but this is all going away as we migrate away from `readLocation` altogether.

Reply:
Makes sense, thanks for the explanation!